### PR TITLE
Use jsoniter to encode object

### PR DESCRIPTION
**Pull request by ggaaooppeenngg**
standard encoding/json has performance issue,  use jsoniter
can help to reduce half of the encoding time.

Signed-off-by: Peng Gao <peng.gao.dut@gmail.com>

**What this PR does / why we need it**:
This PR change encoding/json to jsoniter, from the profiling it reduce the half of processing time.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```
/sig api-machinery

### CODE REVIEW

**Comment by mkumatag** [comment]
I just see only [this](https://github.com/ggaaooppeenngg/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L76) entry in the whole file using `encoding/json`. I think we can use jsoniter to remove the complete dependency of `encoding/json` in this file.

**Comment by ggaaooppeenngg** [comment]
I think it's ok to remove `encoding/json` dependency in this file,` json.Number` underling type is string.

**Comment by liggitt** [comment]
it doesn't appear that jsoniter normalizes key order:
```
--- FAIL: TestDeepCopyApiObjects (3.32s)
	copy_test.go:74: {"name":"iÃ¥Ç©Ãæ¹¶&Å¬;ç¨«nTå’œpç‘å®É‡","generateName":"cÃ“Ã¤É‰ç¤ˆÃ—Ä’Pç¨Šé©É”","selfLink":";c","uid":"Å³ÉƒÇé«€_Kè€‹é¡¬éœ§(-+","resourceVersion":"9655338397498521902","creationTimestamp":null,"Spec":{"AccessModes":["Ä¿vé‡‰ç»`SçŸ˜èƒ„"],"Selector":{"matchLabels":{}},"Resources":{"Limits":{"
		
		A: storage":"28","cpu":"88","memory":"63"},"Requests":{"cpu":"88","memory":"63","storage":"28"}},"VolumeName":"å§¥HÊ‹éŸ›åª æ±›6é…™T'","StorageClassName":""},"Status":{"Phase":"Lost","AccessModes":["Æ‰#å¨ˆç ¿Vè§»zå‘•Â¥Ã¥JÅ¡AIÆŒpÅ¡é‡‘Å›"],"Capacity":null,"Conditions":[]}}
		
		B: cpu":"88","memory":"63","storage":"28"},"Requests":{"memory":"63","storage":"28","cpu":"88"}},"VolumeName":"å§¥HÊ‹éŸ›åª æ±›6é…™T'","StorageClassName":""},"Status":{"Phase":"Lost","AccessModes":["Æ‰#å¨ˆç ¿Vè§»zå‘•Â¥Ã¥JÅ¡AIÆŒpÅ¡é‡‘Å›"],"Capacity":null,"Conditions":[]}}
```

stable order is required

**Comment by liggitt** [comment]
@thockin did you exercise encoding in jsoniter as thoroughly as decoding?

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 1]
```diff
@@ -216,7 +215,7 @@ func (s *Serializer) Encode(obj runtime.Object, w io.Writer) error {
 		_, err = w.Write(data)
 		return err
 	}
-	encoder := json.NewEncoder(w)
+	encoder := jsoniter.ConfigCompatibleWithStandardLibrary.NewEncoder(w)
```
Comment (thockin) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: Everything looks OK except this.  Doesn't this have a negative impact of performance?  It did when I measured it.

Comment (ggaaooppeenngg) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: I offered a bench commit, on my machine it's
```
pkg: k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json
BenchmarkJsonEncoding-4       	 3000000	       569 ns/op
BenchmarkJsoniterEncoding-4   	 5000000	       316 ns/op
```
How did you measure it?

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: The benchmark doesn't use data that is similar to actual objects; can you get e.g. a big pod spec or something and run again?

Comment (ggaaooppeenngg) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: @lavalamp  I pushed a commit, the result was not better, but just 20% improved.
```
pkg: k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json
BenchmarkJsonEncoding-4       	  100000	     13080 ns/op
BenchmarkJsoniterEncoding-4   	  200000	     11158 ns/op
PASS
```

**Review by ggaaooppeenngg — COMMENTED** [review]
I made up a large pod for benchmark.

**Comment by liggitt** [comment]
consistent failure:
```
--- FAIL: TestGetJobFromTemplate (0.00s)
	utils_test.go:96: Wrong length for created-by annotation, expected 160 got 159
	utils_test.go:99: Wrong value for created-by annotation, expected {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"CronJob","namespace":"snazzycats","name":"mycronjob","uid":"1a2b3c","apiVersion":"batch"}}
		 got {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"CronJob","namespace":"snazzycats","name":"mycronjob","uid":"1a2b3c","apiVersion":"batch"}}
```

**Comment by ggaaooppeenngg** [comment]
I made fix https://github.com/json-iterator/go/pull/188 for jsoniter, wait for the test to pass.

**Comment by ggaaooppeenngg** [comment]
Fix non-nil empty map not ignored, https://github.com/json-iterator/go/pull/189

**Comment by ggaaooppeenngg** [comment]
In order to make json-iter to be compatible with standard library, it needs some work, wait for the lib      stable then open it again.

**Pull request closed without merging**